File: Extractor_Agent/server/utils/agent_result.ts
import { inferCrawlerFailureTags, inferCrawlerRouteSuggestion } from "./crawl_route_suggestion";

export type AgentSource = {
  type: "url" | "doc" | "table" | "sql";
  ref: string;
};

export type AgentResult = {
  ok: boolean;
  agent: string;
  trace_id?: string;
  answer?: string;
  sources?: AgentSource[];
  structured?: Record<string, unknown>;
  needs_clarify?: boolean;
  clarify_questions?: string[];
  error_code?: string;
  latency_ms?: number;
};

type CrawlItem = Record<string, unknown>;

export function buildCrawlerAgentResult(params: {
  items?: CrawlItem[];
  outputContent?: string;
  status?: string;
  trace_id?: string;
  latency_ms?: number;
  serp_fallback?: boolean;
  clarifyReason?: string;
  stats?: Record<string, unknown>;
  taskPlan?: { needsLogin?: boolean; preferredChannel?: string };
  lastError?: string;
  meta?: Record<string, unknown>;
  failureTags?: string[];
  planNeedsLogin?: boolean;
}): AgentResult {
  const sources: AgentSource[] = [];
  const seen = new Set<string>();
  for (const item of params.items || []) {
    const url = String(item?.url ?? "").trim();
    if (!url || seen.has(url)) continue;
    seen.add(url);
    sources.push({ type: "url", ref: url });
  }
  const answer = String(params.outputContent || "").trim();
  const itemCount = params.items?.length ?? 0;
  const routeSuggestion = inferCrawlerRouteSuggestion({
    status: params.status,
    clarifyReason: params.clarifyReason,
    itemCount,
    stats: params.stats,
    taskPlan: params.taskPlan,
    lastError: params.lastError,
    failureTags: params.failureTags ?? (Array.isArray(params.meta?.failure_tags) ? (params.meta!.failure_tags as string[]) : undefined),
    planNeedsLogin: params.planNeedsLogin,
  });
  const failureTags = inferCrawlerFailureTags({
    status: params.status,
    clarifyReason: params.clarifyReason,
    itemCount,
    stats: params.stats,
    taskPlan: params.taskPlan,
    lastError: params.lastError,
    failureTags: params.failureTags,
    planNeedsLogin: params.planNeedsLogin,
  });
  const ok =
    String(params.status || "").toLowerCase() === "ok" ||
    sources.length > 0 ||
    Boolean(answer);
  const needsClarify =
    String(params.status || "").toLowerCase() === "needs_clarification" ||
    (!ok && itemCount === 0 && !answer);
  return {
    ok,
    agent: "crawler",
    trace_id: params.trace_id,
    answer: answer || undefined,
    sources: sources.length ? sources : undefined,
    structured: {
      content_trust: "untrusted",
      content_trust_source: "crawler",
      itemCount,
      status: params.status,
      ...(params.serp_fallback ? { serp_fallback: true } : {}),
      ...(params.meta?.seed_first ? { seed_first: true } : {}),
      ...(params.meta?.manager_seed_count != null
        ? { manager_seed_count: params.meta.manager_seed_count }
        : {}),
      ...(params.meta?.cloud_scrape_calls != null
        ? { cloud_scrape_calls: params.meta.cloud_scrape_calls }
        : {}),
      ...(params.meta?.extract_path ? { extract_path: params.meta.extract_path } : {}),
      ...(params.meta?.llm_extract_calls != null ? { llm_extract_calls: params.meta.llm_extract_calls } : {}),
      ...(params.meta?.template_hit ? { template_hit: true } : {}),
      ...(params.meta?.patch_hit ? { patch_hit: true, patch_id: params.meta.patch_id } : {}),
      ...(routeSuggestion ? { route_suggestion: routeSuggestion } : {}),
      ...(failureTags.length ? { failure_tags: failureTags } : {}),
      ...(params.stats?.cloud_scrape_calls != null
        ? { cloud_scrape_calls: params.stats.cloud_scrape_calls }
        : {}),
      ...(params.stats?.manager_seed_count != null
        ? { manager_seed_count: params.stats.manager_seed_count }
        : {}),
      preview: (params.items || []).slice(0, 3).map((it) => ({
        title: String(it?.title ?? "").slice(0, 120),
        url: String(it?.url ?? "").slice(0, 512),
        excerpt: String(it?.excerpt ?? it?.summary ?? "").slice(0, 200),
      })),
    },
    needs_clarify: needsClarify || undefined,
    error_code: ok ? undefined : routeSuggestion === "gui" ? "route_gui_suggested" : "empty_result",
    latency_ms: params.latency_ms,
  };
}

File: ExtractorPy_Agent/app/protocol/agent_result.py
"""Manager-compatible agentResult envelope (agent=crawler)."""

from __future__ import annotations

from typing import Any, TypedDict


class AgentSource(TypedDict):
    type: str
    ref: str


class AgentResult(TypedDict, total=False):
    ok: bool
    agent: str
    trace_id: str
    answer: str
    sources: list[AgentSource]
    structured: dict[str, Any]
    needs_clarify: bool
    clarify_questions: list[str]
    error_code: str
    latency_ms: int


def build_crawler_agent_result(
    *,
    items: list[dict[str, Any]] | None = None,
    output_content: str = "",
    status: str = "ok",
    trace_id: str | None = None,
    latency_ms: int | None = None,
    serp_fallback: bool = False,
    meta: dict[str, Any] | None = None,
    failure_tags: list[str] | None = None,
    error_code: str | None = None,
    needs_clarify: bool = False,
) -> AgentResult:
    items = items or []
    sources: list[AgentSource] = []
    seen: set[str] = set()
    for it in items:
        url = str(it.get("url") or "").strip()
        if not url or url in seen:
            continue
        seen.add(url)
        sources.append({"type": "url", "ref": url})

    answer = (output_content or "").strip()
    item_count = len(items)
    status_l = (status or "").lower()
    ok = status_l == "ok" or bool(sources) or bool(answer)
    if needs_clarify is False and status_l == "needs_clarification":
        needs_clarify = True
    if not ok and item_count == 0 and not answer:
        needs_clarify = True

    structured: dict[str, Any] = {
        "content_trust": "untrusted",
        "content_trust_source": "crawler",
        "itemCount": item_count,
        "status": status,
    }
    meta = meta or {}
    if serp_fallback:
        structured["serp_fallback"] = True
    for k in (
        "seed_first",
        "manager_seed_count",
        "cloud_scrape_calls",
        "playwright_mcp_calls",
        "extract_path",
        "llm_extract_calls",
    ):
        if meta.get(k) is not None:
            structured[k] = meta[k]
    if failure_tags:
        structured["failure_tags"] = failure_tags

    result: AgentResult = {
        "ok": ok,
        "agent": "crawler",
        "structured": structured,
    }
    if trace_id:
        result["trace_id"] = trace_id
    if answer:
        result["answer"] = answer
    if sources:
        result["sources"] = sources
    if needs_clarify:
        result["needs_clarify"] = True
    if latency_ms is not None:
        result["latency_ms"] = latency_ms
    if not ok:
        result["error_code"] = error_code or "CRAWLER_EMPTY"
    return result
